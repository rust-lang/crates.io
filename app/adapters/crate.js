import ApplicationAdapter from './application';

const BULK_REQUEST_GROUP_SIZE = 10;

export default class CrateAdapter extends ApplicationAdapter {
  coalesceFindRequests = true;

  findRecord(store, type, id, snapshot) {
    return super.findRecord(store, type, id, setDefaultInclude(snapshot));
  }

  queryRecord(store, type, query, adapterOptions) {
    return super.queryRecord(store, type, setDefaultInclude(query), adapterOptions);
  }

  /** Removes the `name` query parameter and turns it into a path parameter instead */
  urlForQueryRecord(query) {
    let baseUrl = super.urlForQueryRecord(...arguments);
    if (!query.name) {
      return baseUrl;
    }

    let crateName = query.name;
    delete query.name;
    return `${baseUrl}/${crateName}`;
  }

  /** Adds a `message` query parameter to the URL, if set in the `adapterOptions`. */
  urlForDeleteRecord(id, modelName, snapshot) {
    let url = super.urlForDeleteRecord(...arguments);

    let message = snapshot.adapterOptions.message;
    if (message) {
      url += `?message=${encodeURIComponent(message)}`;
    }

    return url;
  }

  groupRecordsForFindMany(store, snapshots) {
    let result = [];
    for (let i = 0; i < snapshots.length; i += BULK_REQUEST_GROUP_SIZE) {
      result.push(snapshots.slice(i, i + BULK_REQUEST_GROUP_SIZE));
    }
    return result;
  }
}

function setDefaultInclude(query) {
  if (query.include === undefined) {
    // This ensures `crate.versions` are always fetched from another request.
    query.include = 'keywords,categories,downloads';
  }

  return query;
}
