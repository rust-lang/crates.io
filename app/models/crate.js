import Model, { attr, hasMany } from '@ember-data/model';
import { assert } from '@ember/debug';
import { waitForPromise } from '@ember/test-waiters';
import { cached } from '@glimmer/tracking';

import { apiAction } from '@mainmatter/ember-api-actions';
import { task } from 'ember-concurrency';

export default class Crate extends Model {
  @attr name;
  @attr downloads;
  @attr recent_downloads;
  @attr('date') created_at;
  @attr('date') updated_at;
  /**
   * This is the default version that will be shown when visiting the crate
   * details page. Note that this value can be `null`, which may be unexpected.
   * @type {string | null}
   */
  @attr default_version;
  @attr yanked;
  @attr max_version;
  @attr max_stable_version;
  @attr newest_version;

  @attr description;
  @attr homepage;
  @attr documentation;
  @attr repository;

  @hasMany('version', { async: true, inverse: 'crate' }) versions;
  @hasMany('team', { async: true, inverse: null }) owner_team;
  @hasMany('user', { async: true, inverse: null }) owner_user;
  @hasMany('version-download', { async: true, inverse: null }) version_downloads;
  @hasMany('keyword', { async: true, inverse: null }) keywords;
  @hasMany('category', { async: true, inverse: null }) categories;
  @hasMany('dependency', { async: true, inverse: null }) reverse_dependencies;

  @cached get versionIdsBySemver() {
    let { last } = this.loadVersionsTask;
    assert('`loadVersionsTask.perform()` must be called before calling `versionIdsBySemver`', last != null);
    let versions = last?.value ?? [];
    return versions
      .slice()
      .sort(compareVersionBySemver)
      .map(v => v.id);
  }

  @cached get versionIdsByDate() {
    let { last } = this.loadVersionsTask;
    assert('`loadVersionsTask.perform()` must be called before calling `versionIdsByDate`', last != null);
    let versions = last?.value ?? [];
    return versions
      .slice()
      .sort(compareVersionByDate)
      .map(v => v.id);
  }

  @cached get firstVersionId() {
    return this.versionIdsByDate.at(-1);
  }

  @cached get versionsObj() {
    let { last } = this.loadVersionsTask;
    assert('`loadVersionsTask.perform()` must be called before calling `versionsObj`', last != null);
    let versions = last?.value ?? [];
    return Object.fromEntries(versions.slice().map(v => [v.id, v]));
  }

  @cached get releaseTrackSet() {
    let map = new Map();
    let { versionsObj: versions, versionIdsBySemver } = this;
    for (let id of versionIdsBySemver) {
      let { releaseTrack, isPrerelease, yanked } = versions[id];
      if (releaseTrack && !isPrerelease && !yanked && !map.has(releaseTrack)) {
        map.set(releaseTrack, id);
      }
    }
    return new Set(map.values());
  }

  hasOwnerUser(userId) {
    let { last } = this.loadOwnerUserTask;
    assert('`loadOwnerUserTask.perform()` must be called before calling `hasOwnerUser()`', last != null);
    return (last?.value ?? []).some(({ id }) => id === userId);
  }

  get owners() {
    let { last } = this.loadOwnersTask;
    assert('`loadOwnersTask.perform()` must be called before accessing `owners`', last != null);
    return last?.value ?? [];
  }

  async follow() {
    return await waitForPromise(apiAction(this, { method: 'PUT', path: 'follow' }));
  }

  async unfollow() {
    return await waitForPromise(apiAction(this, { method: 'DELETE', path: 'follow' }));
  }

  async inviteOwner(username) {
    let response = await waitForPromise(
      apiAction(this, { method: 'PUT', path: 'owners', data: { owners: [username] } }),
    );
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }

  async removeOwner(username) {
    let response = await waitForPromise(
      apiAction(this, { method: 'DELETE', path: 'owners', data: { owners: [username] } }),
    );
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }

  loadOwnerUserTask = task(async () => {
    return (await this.owner_user) ?? [];
  });

  loadOwnersTask = task(async () => {
    let [teams, users] = await Promise.all([this.owner_team, this.owner_user]);
    return [...(teams ?? []), ...(users ?? [])];
  });

  loadVersionsTask = task(async ({ reload = false } = {}) => {
    let versionsRef = this.hasMany('versions');
    let fut = reload === true ? versionsRef.reload() : versionsRef.load();
    return (await fut) ?? [];
  });
}

function compareVersionBySemver(a, b) {
  let aSemver = a.semver;
  let bSemver = b.semver;

  if (aSemver === bSemver) {
    return b.created_at - a.created_at;
  } else if (aSemver === null) {
    return 1;
  } else if (bSemver === null) {
    return -1;
  } else {
    return bSemver.compare(aSemver);
  }
}

function compareVersionByDate(a, b) {
  let bDate = b.created_at.getTime();
  let aDate = a.created_at.getTime();

  return bDate === aDate ? parseInt(b.id) - parseInt(a.id) : bDate - aDate;
}
