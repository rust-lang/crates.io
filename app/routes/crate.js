import { NotFoundError } from '@ember-data/adapter/error';
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';
import { waitForPromise } from '@ember/test-waiters';

export default class CrateRoute extends Route {
  @service headData;
  @service router;
  @service store;

  async model(params, transition) {
    let crateName = params.crate_id;

    try {
      // We would like the peeked crate to include information (such as keywords) for further
      // processing. Currently, we determine this by checking if associated versions exist,
      // as default_version is included in the queryRecord call.
      // See: https://github.com/rust-lang/crates.io/issues/10663
      let crate = this.store.peekRecord('crate', crateName);
      if (!crate || crate.hasMany('versions').value() == null) {
        crate = await this.store.queryRecord('crate', { name: crateName });
      }
      return crate;
    } catch (error) {
      if (error instanceof NotFoundError) {
        let title = `${crateName}: Crate not found`;
        this.router.replaceWith('catch-all', { transition, error, title });
      } else {
        let title = `${crateName}: Failed to load crate data`;
        this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
      }
    }
  }

  setupController(controller, model) {
    super.setupController(...arguments);
    this.headData.crate = model;
    waitForPromise(model.loadOwnerUserTask.perform()).catch(() => {
      // ignore all errors if the request fails
    });
  }

  resetController() {
    super.resetController(...arguments);
    this.headData.crate = null;
  }
}
