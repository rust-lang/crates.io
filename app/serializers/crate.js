import ApplicationSerializer from './application';

const SKIP_NULL_FIELDS = new Set(['categories', 'keywords']);

export default class CrateSerializer extends ApplicationSerializer {
  isNewSerializerAPI = true;

  extractRelationships(modelClass, resourceHash) {
    if (resourceHash.versions == null) {
      delete resourceHash.versions;
    }

    return super.extractRelationships(...arguments);
  }

  normalizeQueryResponse(_store, _modelClass, payload) {
    // We don't want existing relationships overwritten by results with null values.
    // See: https://github.com/rust-lang/crates.io/issues/10711
    if (payload.crates) {
      payload.crates = payload.crates.map(crate => {
        for (const rel of SKIP_NULL_FIELDS) {
          if (crate[rel] === null) {
            delete crate[rel];
          }
        }
        return crate;
      });
    }
    return super.normalizeQueryResponse(...arguments);
  }
}
