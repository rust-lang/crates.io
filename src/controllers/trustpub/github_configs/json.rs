use axum::Json;
use axum::extract::FromRequest;
use chrono::{DateTime, Utc};

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct GitHubConfig {
    #[schema(example = 42)]
    pub id: i32,
    #[schema(example = "regex")]
    #[serde(rename = "crate")]
    pub krate: String,
    #[schema(example = "rust-lang")]
    pub repository_owner: String,
    #[schema(example = 5430905)]
    pub repository_owner_id: i32,
    #[schema(example = "regex")]
    pub repository_name: String,
    #[schema(example = "ci.yml")]
    pub workflow_filename: String,
    #[schema(example = json!(null))]
    pub environment: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
pub struct NewGitHubConfig {
    #[schema(example = "regex")]
    #[serde(rename = "crate")]
    pub krate: String,
    #[schema(example = "rust-lang")]
    pub repository_owner: String,
    #[schema(example = "regex")]
    pub repository_name: String,
    #[schema(example = "ci.yml")]
    pub workflow_filename: String,
    #[schema(example = json!(null))]
    pub environment: Option<String>,
}

#[derive(Debug, Deserialize, FromRequest, utoipa::ToSchema)]
#[from_request(via(Json))]
pub struct CreateRequest {
    pub github_config: NewGitHubConfig,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct CreateResponse {
    pub github_config: GitHubConfig,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct ListResponse {
    pub github_configs: Vec<GitHubConfig>,
}
