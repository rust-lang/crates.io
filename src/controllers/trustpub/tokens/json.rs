use axum::Json;
use axum::extract::FromRequest;

#[derive(Debug, Deserialize, FromRequest, utoipa::ToSchema)]
#[from_request(via(Json))]
pub struct ExchangeRequest {
    pub jwt: String,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct ExchangeResponse {
    pub token: String,
}
