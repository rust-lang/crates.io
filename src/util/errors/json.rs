use axum::response::{IntoResponse, Response};
use axum::Extension;
use axum_extra::json;
use std::borrow::Cow;
use std::fmt;

use super::{AppError, BoxedAppError};

use crate::middleware::log_request::CauseField;
use crate::rate_limiter::LimitedAction;
use chrono::NaiveDateTime;
use http::{header, StatusCode};

/// Generates a response with the provided status and description as JSON
fn json_error(detail: &str, status: StatusCode) -> Response {
    let json = json!({ "errors": [{ "detail": detail }] });
    (status, json).into_response()
}

// The following structs wrap owned data and provide a custom message to the user

pub fn custom(status: StatusCode, detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    Box::new(CustomApiError {
        status,
        detail: detail.into(),
    })
}

#[derive(Debug, Clone)]
pub struct CustomApiError {
    status: StatusCode,
    detail: Cow<'static, str>,
}

impl fmt::Display for CustomApiError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.detail.fmt(f)
    }
}

impl AppError for CustomApiError {
    fn response(&self) -> Response {
        json_error(&self.detail, self.status)
    }
}

#[derive(Debug)]
pub(crate) struct TooManyRequests {
    pub action: LimitedAction,
    pub retry_after: NaiveDateTime,
}

impl AppError for TooManyRequests {
    fn response(&self) -> Response {
        const HTTP_DATE_FORMAT: &str = "%a, %d %b %Y %T GMT";
        let retry_after = self.retry_after.format(HTTP_DATE_FORMAT);

        let detail = format!(
            "{}. Please try again after {retry_after} or email \
             help@crates.io to have your limit increased.",
            self.action.error_message()
        );
        let mut response = json_error(&detail, StatusCode::TOO_MANY_REQUESTS);
        response.headers_mut().insert(
            header::RETRY_AFTER,
            retry_after
                .to_string()
                .try_into()
                .expect("HTTP_DATE_FORMAT contains invalid char"),
        );
        response
    }
}

impl fmt::Display for TooManyRequests {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        "Too many requests".fmt(f)
    }
}

#[derive(Debug, Clone, Copy)]
pub struct InsecurelyGeneratedTokenRevoked;

impl InsecurelyGeneratedTokenRevoked {
    pub fn boxed() -> BoxedAppError {
        Box::new(Self)
    }
}

impl AppError for InsecurelyGeneratedTokenRevoked {
    fn response(&self) -> Response {
        let cause = CauseField("insecurely generated, revoked 2020-07".to_string());
        let response = json_error(&self.to_string(), StatusCode::UNAUTHORIZED);
        (Extension(cause), response).into_response()
    }
}

pub const TOKEN_FORMAT_ERROR: &str =
    "The given API token does not match the format used by crates.io. \
    \
    Tokens generated before 2020-07-14 were generated with an insecure \
    random number generator, and have been revoked. You can generate a \
    new token at https://crates.io/me. \
    \
    For more information please see \
    https://blog.rust-lang.org/2020/07/14/crates-io-security-advisory.html. \
    We apologize for any inconvenience.";

impl fmt::Display for InsecurelyGeneratedTokenRevoked {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(TOKEN_FORMAT_ERROR)?;
        Result::Ok(())
    }
}
