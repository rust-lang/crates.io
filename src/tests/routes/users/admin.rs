use chrono::{DateTime, Utc};
use crates_io_database::schema::users;
use http::StatusCode;
use insta::{assert_json_snapshot, assert_snapshot};
use serde_json::json;

use crate::{
    models::User,
    tests::util::{RequestHelper, TestApp},
};

mod get {
    use super::*;

    #[tokio::test(flavor = "multi_thread")]
    async fn get() {
        let (app, anon, user) = TestApp::init().with_user().await;
        let admin = app.db_new_admin_user("admin").await;

        // Anonymous users should be forbidden.
        let response = anon.get::<()>("/api/v1/users/foo/admin").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("anonymous-found", response.text());

        let response = anon.get::<()>("/api/v1/users/bar/admin").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("anonymous-not-found", response.text());

        // Regular users should also be forbidden, even if they're requesting
        // themself.
        let response = user.get::<()>("/api/v1/users/foo/admin").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("non-admin-found", response.text());

        let response = user.get::<()>("/api/v1/users/bar/admin").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("non-admin-not-found", response.text());

        // Admin users are allowed, but still can't manifest users who don't
        // exist.
        let response = admin.get::<()>("/api/v1/users/bar/admin").await;
        assert_eq!(response.status(), StatusCode::NOT_FOUND);
        assert_snapshot!("admin-not-found", response.text());

        // Admin users are allowed, and should get an admin's eye view of the
        // requested user.
        let response = admin.get::<()>("/api/v1/users/foo/admin").await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!("admin-found", response.json());
    }
}

mod lock {
    use super::*;

    #[tokio::test(flavor = "multi_thread")]
    async fn lock() {
        let (app, anon, user) = TestApp::init().with_user().await;
        let admin = app.db_new_admin_user("admin").await;

        // Because axum will validate and deserialise the body before any auth
        // check occurs, we actually need to provide a valid body for all the
        // auth related test cases.
        let body = serde_json::to_string(&json!({
            "reason": "l33t h4x0r",
            "until": "2045-01-01T01:02:03Z",
        }))
        .unwrap();

        // Anonymous users should be forbidden.
        let response = anon.put::<()>("/api/v1/users/foo/lock", body.clone()).await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("anonymous-found", response.text());

        let response = anon.put::<()>("/api/v1/users/bar/lock", body.clone()).await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("anonymous-not-found", response.text());

        // Regular users should also be forbidden, even if they're locking
        // themself.
        let response = user.put::<()>("/api/v1/users/foo/lock", body.clone()).await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("non-admin-found", response.text());

        let response = user.put::<()>("/api/v1/users/bar/lock", body.clone()).await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("non-admin-not-found", response.text());

        // Admin users are allowed, but still can't manifest users who don't
        // exist.
        let response = admin
            .put::<()>("/api/v1/users/bar/lock", body.clone())
            .await;
        assert_eq!(response.status(), StatusCode::NOT_FOUND);
        assert_snapshot!("admin-not-found", response.text());

        // Admin users who provide invalid request bodies should be denied.
        let response = admin
            .put::<()>("/api/v1/users/bar/lock", b"invalid JSON" as &[u8])
            .await;
        assert_eq!(response.status(), StatusCode::UNSUPPORTED_MEDIA_TYPE);
        assert_snapshot!("admin-invalid-json", response.text());

        let response = admin
            .put::<()>("/api/v1/users/bar/lock", br#"{"valid": "json"}"# as &[u8])
            .await;
        assert_eq!(response.status(), StatusCode::UNPROCESSABLE_ENTITY);
        assert_snapshot!("admin-malformed-json", response.text());

        // Admin users are allowed, and should be able to lock the user.
        assert_none!(&user.as_model().account_lock_reason);
        assert_none!(&user.as_model().account_lock_until);

        let response = admin.put::<()>("/api/v1/users/foo/lock", body).await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!("admin-found", response.json());

        // Get the user again and validate that they are now locked.
        let mut conn = app.db_conn().await;
        let locked_user = User::find(&mut conn, user.as_model().id).await.unwrap();
        assert_user_is_locked(&locked_user, "l33t h4x0r", "2045-01-01T01:02:03Z");

        // Re-locking a locked user should update their lock reason and
        // expiration time.
        let body = serde_json::to_string(&json!({
            "reason": "less l33t",
            "until": "2035-01-01T01:02:03Z",
        }))
        .unwrap();

        let response = admin.put::<()>("/api/v1/users/foo/lock", body).await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!("admin-relock-shorter", response.json());

        // Get the user again and validate that they are now locked for less
        // time.
        let mut conn = app.db_conn().await;
        let locked_user = User::find(&mut conn, user.as_model().id).await.unwrap();
        assert_user_is_locked(&locked_user, "less l33t", "2035-01-01T01:02:03Z");

        // Finally, not including an until time at all should lock the account
        // forever. (Insert evil laughter here.)
        let body = serde_json::to_string(&json!({
            "reason": "less l33t",
        }))
        .unwrap();

        let response = admin.put::<()>("/api/v1/users/foo/lock", body).await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!("admin-lock-forever", response.json());

        // Get the user again and validate that they are now locked forever.
        let mut conn = app.db_conn().await;
        let locked_user = User::find(&mut conn, user.as_model().id).await.unwrap();
        assert_user_is_locked_indefinitely(&locked_user, "less l33t");
    }
}

mod unlock {
    use super::*;

    #[tokio::test(flavor = "multi_thread")]
    async fn unlock() {
        let (app, anon, user) = TestApp::init().with_user().await;
        let admin = app.db_new_admin_user("admin").await;

        use diesel::prelude::*;
        use diesel_async::RunQueryDsl;

        // First up, let's lock the user.
        let mut conn = app.db_conn().await;
        diesel::update(user.as_model())
            .set((
                users::account_lock_reason.eq("naughty naughty"),
                users::account_lock_until.eq(DateTime::parse_from_rfc3339("2050-01-01T01:02:03Z")
                    .unwrap()
                    .naive_utc()),
            ))
            .execute(&mut conn)
            .await
            .unwrap();

        // Anonymous users should be forbidden.
        let response = anon.delete::<()>("/api/v1/users/foo/lock").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("anonymous-found", response.text());

        let response = anon.delete::<()>("/api/v1/users/bar/lock").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("anonymous-not-found", response.text());

        // Regular users should also be forbidden, even if they're locking
        // themself.
        let response = user.delete::<()>("/api/v1/users/foo/lock").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("non-admin-found", response.text());

        let response = user.delete::<()>("/api/v1/users/bar/lock").await;
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_snapshot!("non-admin-not-found", response.text());

        // Admin users are allowed, but still can't manifest users who don't
        // exist.
        let response = admin.delete::<()>("/api/v1/users/bar/lock").await;
        assert_eq!(response.status(), StatusCode::NOT_FOUND);
        assert_snapshot!("admin-not-found", response.text());

        // Admin users are allowed, and should be able to unlock the user.
        let response = admin.delete::<()>("/api/v1/users/foo/lock").await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!("admin-found", response.json(), {
            ".lock.until" => "[datetime]",
        });

        // Get the user again and validate that they are now unlocked.
        let mut conn = app.db_conn().await;
        let unlocked_user = User::find(&mut conn, user.as_model().id).await.unwrap();
        assert_user_is_unlocked(&unlocked_user);

        // Unlocking an unlocked user should succeed silently.
        let response = admin.delete::<()>("/api/v1/users/foo/lock").await;
        assert_eq!(response.status(), StatusCode::OK);
        assert_json_snapshot!("admin-reunlock", response.json(), {
            ".lock.until" => "[datetime]",
        });
    }
}

#[track_caller]
fn assert_user_is_locked(user: &User, reason: &str, until: &str) {
    assert_eq!(user.account_lock_reason.as_deref(), Some(reason));
    assert_eq!(
        user.account_lock_until,
        Some(DateTime::parse_from_rfc3339(until).unwrap().naive_utc())
    );
}

#[track_caller]
fn assert_user_is_locked_indefinitely(user: &User, reason: &str) {
    assert_eq!(user.account_lock_reason.as_deref(), Some(reason));
    assert_none!(user.account_lock_until);
}

#[track_caller]
fn assert_user_is_unlocked(user: &User) {
    if user.account_lock_reason.is_some() {
        if let Some(until) = user.account_lock_until {
            assert_lt!(until, Utc::now().naive_utc());
        } else {
            panic!("user account is locked indefinitely");
        }
    }
}
