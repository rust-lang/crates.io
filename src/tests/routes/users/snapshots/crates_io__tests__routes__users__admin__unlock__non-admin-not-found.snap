---
source: src/tests/routes/users/admin.rs
expression: response.text()
---
{"errors":[{"detail":"This account is locked until 2050-01-01 at 01:02:03 UTC. Reason: naughty naughty"}]}
